### PR TITLE
[SPORK-69] Aspiration windows

### PR DESCRIPTION
**Pull request by KYLChiu**
### Notice in most of these the aspiration window performs equal/better to base case. 

**Position: [rnbqkb1r/pppp1ppp/4pn2/8/3P1B2/8/PPP1PPPP/RN1QKBNR w KQkq - 0 3]**
test_perf_base -  65512033 function calls (65346004 primitive calls) in 23.012 seconds
test_perf_null_move_pruning -  54089629 function calls (53950069 primitive calls) in 19.221 seconds
test_perf_aspiration_windows -  48245399 function calls (48126490 primitive calls) in 17.999 seconds
test_perf_combined -  37158849 function calls (37064983 primitive calls) in 13.256 seconds

**Position - [r1b2rk1/ppq2ppp/3bpn2/3pP3/8/2PB2B1/PP1N1PPP/R2QK2R b KQ - 0 11]**
test_perf_base          86977875 function calls (86772553 primitive calls) in 29.970 seconds
test_perf_null_move_pruning      30151924 function calls (30073541 primitive calls) in 10.358 seconds
test_perf_aspiration_windows         88133553 function calls (87927314 primitive calls) in 30.902 seconds
test_perf_combined    29851977 function calls (29775617 primitive calls) in 10.616 seconds

**Position - [8/8/8/8/5R2/2pk4/5K2/8 b - - 0 1]**
:test_perf_base   20941366 function calls (20868306 primitive calls) in 6.527 seconds
test_perf_null_move_pruning      13473859 function calls (13425207 primitive calls) in 4.122 seconds
test_perf_aspiration_windows 9541708 function calls (9510852 primitive calls) in 3.144 seconds
test_perf_combined          16163313 function calls (16105750 primitive calls) in 5.176 seconds

### However sometimes we will have to do extra work:

**Position: [r1b2rk1/ppqn1pbp/6p1/3pp3/7B/2PB1N2/PP3PPP/R2QR1K1 w - - 0 16]**
test_perf_base -  51990918 function calls (51883345 primitive calls) in 17.879 seconds
:test_perf_null_move_pruning - 47226617 function calls (47128862 primitive calls) in 16.379 seconds
test_perf_aspiration_windows - 87418741 function calls (87242088 primitive calls) in 31.540 seconds
test_perf_combined -  79167462 function calls (79007405 primitive calls) in 27.374 seconds

### CODE REVIEW

File: sporkfish/searcher.py [hunk 1]
```diff
@@ -348,6 +350,59 @@ def _negamax_lazy_smp(
...
+    # Assumption (*): we expect the search score from depth n should be somewhat similar depth n + 1.
+    # Method: set alpha = prev_depth_score - constant, beta = prev_depth_score + constant.
+    # Do a search with narrower window [alpha, beta] as opposed to [-inf, inf].
+    # This should cause cutoffs (if they do happen) to happen faster.
+    # If the score falls within the alpha/beta bounds, our assumption (*) holds - so take that score.
+    # If the score falls outside the alpha/beta windows, the score changed too much
+    # i.e. our assumption (*) is false - so we do a full search to be safe.
+    def _aspiration_windows_search(
+        self,
+        search_func: Callable[[Board, int, float, float], Tuple[float, chess.Move]],
```
Comment (ccjeremylo) [inline sporkfish/searcher.py:363]: Wasn't immediately clear to me what the `search_func` mean. I can see it is calling, e.g. `negamax_sp`.

I think would be clearer when I refactored this searcher class a bit.

File: tests/test_searcher.py [hunk 2]
```diff
@@ -49,7 +51,10 @@ def test_valid_moves(self, fen_string: str):
 @pytest.mark.parametrize(
     ("fen_string", "max_depth"),
     [
-        ("r1r3k1/1ppp1ppp/p7/8/1P1nPPn1/3B1RP1/P1PP3q/R1BQ2K1 w - - 2 18", 6),
+        (board_setup["white"]["mid"], 4),
```
Comment (ccjeremylo) [inline tests/test_searcher.py:54]: nice one - thanks for tidying this up

File: sporkfish/searcher.py [hunk 3]
```diff
@@ -370,15 +425,19 @@ def do_search_with_info(
             board_to_search: Board,
             depth: int,
             start_time: float,
-            alpha: float,
-            beta: float,
+            prev_score: float,
```
Comment (ccjeremylo) [inline sporkfish/searcher.py:428]: I prefer this interface now

File: sporkfish/searcher.py [hunk 4]
```diff
@@ -411,12 +474,13 @@ def do_search_with_info(
         move = chess.Move.null()
 
         # Iterative deepening
+        # This conducts a fixed-depth search for depths ranging from 1 to max_depth.
```
Comment (ccjeremylo) [inline sporkfish/searcher.py:477]: nice comment - suggest in the future, when we have wiki pages - we can just change the comments to point to corresponding wiki pages.

File: sporkfish/searcher.py [hunk 5]
```diff
@@ -348,6 +350,59 @@ def _negamax_lazy_smp(
...
+        :param prev_score: The score of the previous depth in a iterative deepening search.
+        :type prev_score: float
+
+        :return: A tuple containing the score and the best move found during the search.
+        :rtype: Tuple[float, chess.Move]
+        """
+        if self._config.enable_aspiration_windows and depth > 1:
+            # Aspiration window size: 50 centipawn is worth about 1/2 of a pawn in our eval function
+            # We leave configuration for this to another PR
+            window_size = 50
```
Comment (ccjeremylo) [inline sporkfish/searcher.py:389]: According to your [reference](https://www.chessprogramming.org/Aspiration_Windows) - "Typical window sizes are 1/2 to 1/4 of a pawn on either side of the guess.".

I am struggling to understand how that corresponds to the `window_size` of 50 you are using here. Can you elaborate the unit here?

[edit] Clearly I was blind, just read your comment. All clear thanks

**Review by ccjeremylo — APPROVED** [review]
Interesting stuff, I really like the documentation on perf analysis. Not surprising that sometimes we get fast search, sometimes slower.

I think good idea to leave the config as on as of now.

Left some comments, but overall LGTM. Approved!

File: sporkfish/searcher.py (hunk 1 above)
Comment (KYLChiu) [reply sporkfish/searcher.py:363]: Yes - hopefully abit clearer in MiniMaxVariant, which can directly call the search function.

File: tests/test_searcher.py [hunk 6]
```diff
@@ -105,6 +115,25 @@ def test_perf_null_move_pruning(self, fen_string: str, max_depth: int) -> None:
             enable_null_move_pruning=True,
         )
 
+    @pytest.mark.slow
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply tests/test_searcher.py:118]: Sounds good - will add to next PR (if someone hasn't done already).

**Pull request merged**